src/assignments/yunfeng/lesson_4_answers.clj

```diff
@@ -0,0 +1,138 @@
+(remove-duplicate-ranks [{:suit :spade :rank 12} {:suit :club :rank 12} {:suit :heart :rank 10} {:suit :diamond :rank 11} {:suit :spade :rank 3}])
+(remove-duplicate-ranks [{:suit :spade :rank 12} {:suit :club :rank 9} {:suit :heart :rank 10} {:suit :diamond :rank 11} {:suit :spade :rank 3}])
+
+
+(defn find-duplicate-ranks
+  "It is trying to find duplicate ranks regardless of what the suit is"
+  [recongized-hand]
+  (->> (map :rank recongized-hand)
+       (frequencies)
+       (filter #(-> % val (> 1)))
+       (keys)))
+
+;;test find-duplicate-ranks function
+(find-duplicate-ranks [{:suit :spade :rank 12} {:suit :spade :rank 0} {:suit :spade :rank 1} {:suit :spade :rank 2} {:suit :spade :rank 3}])
+(find-duplicate-ranks [{:suit :diamond :rank 0} {:suit :spade :rank 0} {:suit :spade :rank 1} {:suit :spade :rank 2} {:suit :spade :rank 3}])
+
+
+(defn flush?
+  "Check if a hand is a flush"
+  [recongized-hand]
+  (->> (map :suit recongized-hand)
+       (set)
+       (count)
+       (= 1)))
+
+;;test flush? function
+(flush? [{:suit :spade :rank 12} {:suit :spade :rank 0} {:suit :spade :rank 1} {:suit :spade :rank 2} {:suit :spade :rank 3}])
+(flush? [{:suit :club :rank 12} {:suit :spade :rank 0} {:suit :spade :rank 1} {:suit :spade :rank 2} {:suit :spade :rank 3}])
+
+
+(defn straight?
+  "Check if a hand is a straight
+   A 2 3 4 5 is a special case"
+  [recongized-hand]
+  (let [sorted-rank (sort (map :rank recongized-hand))]
+    (or (= 4 (- (last sorted-rank) (first sorted-rank))) (= [0 1 2 3 12] sorted-rank))))
+
+;;test straight? function
+(straight? [{:suit :spade :rank 12} {:suit :spade :rank 0} {:suit :spade :rank 1} {:suit :spade :rank 2} {:suit :spade :rank 3}])
+(straight? [{:suit :spade :rank 4} {:suit :spade :rank 0} {:suit :spade :rank 1} {:suit :spade :rank 2} {:suit :spade :rank 3}])
+(straight? [{:suit :spade :rank 5} {:suit :spade :rank 0} {:suit :spade :rank 1} {:suit :spade :rank 2} {:suit :spade :rank 3}])
+
+
+(defn evaluate-hand
+  "It uses the number of remaining ranks after removing all the duplicate ranks in a hand to seprate cases. 
+   case 1:  If there are 5 remaining ranks(no two ranks are the same) which means this hand cannot be pair, 2 pair, 3 of a kind, 4 of a kind or full house.
+            Then we can check if the hand is a flush or straight to determine the final result.
+   case 2:  If there are 4 remaining ranks(1 duplicate ranks) which means this hand can only be pair.
+   case 3:  If there are 3 remaining ranks(2 duplicate ranks) which means this hand can only be two pair or 3 of a kind.
+            Then we can check if the not unique ranks are all the same to determine the result.
+   case 4:  If there are 2 remaining ranks(3 duplicate ranks) which means this hand can only be full house or 4 of a kind.
+            We use the same method as case 3 to determine the final result.
+   default: means invalid input"
+  [recongized-hand number-of-unique-rank]
+  (case number-of-unique-rank
```

Comment:
All of these nested case/if statements make things very scary looking.  There is a way to express this very cleanly.

Also, look into the COND macro.
Cheers!

Reply:
Yeah sure. I chose to use case and if statements because there are 9 possible cases. I will try to find a  better way to do this.

Reply:
It looks better when I use cond function instead of nested case/if statement. Thanks